ivadomed/loader/slice_filter.py

```diff
@@ -4,65 +4,101 @@
 
-        if self.filter_empty_mask:
-            # Filter slices that do not have ANY ground truth (i.e. all masks are empty)
-            if not np.any(gt_data):
-                return False
+        if is_2d_patch:
+            if self.is_train:
+                if self.filter_empty_mask_patch:
+                    # Filter 2D patches at training time that do not have ANY ground truth (i.e. all masks are empty)
+                    if not np.any(gt_data):
+                        return False
 
-        if self.filter_absent_class:
-            # Filter slices that have absent classes (i.e. one or more masks are empty)
-            if not np.all([np.any(mask) for mask in gt_data]):
-                return False
+                if self.filter_absent_class_patch:
+                    # Filter 2D patches at training time that have absent classes (i.e. one or more masks are empty)
+                    if not np.all([np.any(mask) for mask in gt_data]):
+                        return False
 
-        if self.filter_empty_input:
-            # Filter set of images if one of them is empty or filled with constant value (i.e. std == 0)
-            if np.any([img.std() == 0 for img in input_data]):
-                return False
+                if self.filter_empty_input_patch:
+                    # Filter set of 2D patches at training time if one of them is empty or filled with constant value (i.e. std == 0)
+                    if np.any([img.std() == 0 for img in input_data]):
+                        return False
+        else:
+            if self.filter_empty_mask:
+                # Filter slices that do not have ANY ground truth (i.e. all masks are empty)
+                if not np.any(gt_data):
+                    return False
 
-        if self.filter_classification:
-            if not np.all([int(
-                    self.classifier(
-                        imed_utils.cuda(torch.from_numpy(img.copy()).unsqueeze(0).unsqueeze(0),
-                                        self.cuda_available))) for img in input_data]):
-                return False
+            if self.filter_absent_class:
+                # Filter slices that have absent classes (i.e. one or more masks are empty)
+                if not np.all([np.any(mask) for mask in gt_data]):
+                    return False
+
+            if self.filter_empty_input:
+                # Filter set of images if one of them is empty or filled with constant value (i.e. std == 0)
+                if np.any([img.std() == 0 for img in input_data]):
+                    return False
+
+            if self.filter_classification:
+                if not np.all([int(
+                        self.classifier(
+                            imed_utils.cuda(torch.from_numpy(img.copy()).unsqueeze(0).unsqueeze(0),
+                                            self.cuda_available))) for img in input_data]):
+                    return False
 
         return True
```

Comment:
Overall, I think these parts needs to be simplified. 

Multilayered `if`s are typically error prone. Same logic can be expressed as shown below: 

```
# Filter 2D patches at training time that 
if is_2d_patch and self.is_train:
    # ... do not have ANY ground truth (i.e. all masks are empty)
    if self.filter_empty_mask_patch and not np.any(gt_data):
        return False

    # ... have absent classes (i.e. one or more masks are empty)
    if self.filter_absent_class_patch and not np.all([np.any(mask) for mask in gt_data]):
        return False
else:

  # Filter slices that do not have ANY ground truth (i.e. all masks are empty)
  if self.filter_empty_mask and not np.any(gt_data):
      return False

  # Filter slices that have absent classes (i.e. one or more masks are empty)      
  if self.filter_absent_class and not np.all([np.any(mask) for mask in gt_data]):
      return False

  # Filter set of images if one of them is empty or filled with constant value (i.e. std == 0)      
  if self.filter_empty_input and np.any([img.std() == 0 for img in input_data]):
      return False

  # Filter unmatched class? 
  matching_class = np.all([int(self.classifier(
                  imed_utils.cuda(torch.from_numpy(img.copy()).unsqueeze(0).unsqueeze(0),
                                  self.cuda_available))) for img in input_data])
  if self.filter_classification and not matching_class :    
      return False

return True
```

I also recommend avoid one liners like the `matching_class` boolean assignment I extracted out. IF a one liner like that is needed, there probably should be at least a paragraph explaining what it is doing. Typically it is far more efficient to write it out into smaller (albeit less efficient) statements that make it crystal clear what each steps are doing. It also benefit from type hinting and IDE validation along each small step.

Reply:
Yup. Good point. I think the complexity warrant a split for clarity and you won't have to worry about the complex if conditions here and should be greatly simplified and more readable. Please disregard my overly complicated example lol. 

I will probably add a classifier filter refactor down the road to avoid complexity like that. Again, out of scope of this particular PR, thanks for the clarification.